### PR TITLE
Implement Concurrency For Performance Optimization : RnD

### PR DESCRIPTION
**Pull request by ndegwamartin**
Notes

- Optimizations to Profile page loading
- Optimizations to Register page

**Checklist**
- [ ] I have written **Unit tests** for any new feature(s) and edge cases for bug fixes
- [ ] I have added any strings visible on UI components to the `strings.xml` file
- [ ] I have updated the  [CHANGELOG.md](./CHANGELOG.md) file for any notable changes to the codebase
- [ ] I have run `./gradlew spotlessApply` and `./gradlew spotlessCheck` to check my code follows the project's style guide
- [ ] I have built and run the fhircore app to verify my change fixes the issue and/or does not break the app

### CODE REVIEW

File: android/engine/src/main/java/org/smartregister/fhircore/engine/rulesengine/RulesExecutor.kt [hunk 1]
```diff
@@ -80,26 +81,24 @@ class RulesExecutor @Inject constructor(val rulesFactory: RulesFactory) {
...
   private suspend fun List<Resource>.mapToResourceData(
     relatedResourcesMap: MutableMap<String, MutableList<Resource>>,
     ruleConfigs: List<RuleConfig>,
-    ruleConfigsKey: String,
     listRelatedResources: List<ExtractedResource>,
     computedValuesMap: Map<String, Any>
   ) =
     this.map { resource ->
-      val start = System.currentTimeMillis()
+      val start0 = System.currentTimeMillis()
```
Comment (ellykits) [inline android/engine/src/main/java/org/smartregister/fhircore/engine/rulesengine/RulesExecutor.kt:101]: You can use Kotlin's lambda function [measureTimeMillis](https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.system/measure-time-millis.html)

File: android/engine/src/main/java/org/smartregister/fhircore/engine/rulesengine/RulesExecutor.kt [hunk 2]
```diff
@@ -111,16 +110,20 @@ class RulesExecutor @Inject constructor(val rulesFactory: RulesFactory) {
               relatedResourcesMap = relatedResourcesMap
             )
         }
+      Timber.d(
+        "ListRelatedResources.associate executed in ${ System.currentTimeMillis() - start0 } millisecond(s)"
+      )
 
-      val start2 = System.currentTimeMillis()
-
+      val start = System.currentTimeMillis()
```
Comment (ellykits) [inline android/engine/src/main/java/org/smartregister/fhircore/engine/rulesengine/RulesExecutor.kt:117]: You can use Kotlin's lambda function [measureTimeMillis](https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.system/measure-time-millis.html)

File: android/engine/src/main/java/org/smartregister/fhircore/engine/rulesengine/RulesFactory.kt [hunk 3]
```diff
@@ -150,30 +149,27 @@ constructor(
...
-                }
-              }
-              .toSet()
-          )
-          .also { ruleConfigsCache[ruleConfigsKey] = it }
-      )
-
-    return jexlRules
+  suspend fun generateRules(ruleConfigs: List<RuleConfig>): Rules {
+    val start = System.currentTimeMillis()
```
Comment (ellykits) [inline android/engine/src/main/java/org/smartregister/fhircore/engine/rulesengine/RulesFactory.kt:153]: You can use Kotlin's lambda function [measureTimeMillis](https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.system/measure-time-millis.html)

**Comment by ndegwamartin** [comment]
The user tests for this PR didn’t seem to produce any significant performance improvement for the app. The areas where we would have gained via concurrency the most were in DB interaction however sqlite DB on Android does not support multithreading in a straight forward way. There are some PRAGMA flags that could alter the way it works for us e.g. enabling [Write Ahead Logging](https://sqlite.org/wal.html) but there’s the trade off in that we risk durability i.e. loosing some data incase for instance a battery is removed from the device. On Android reading and writing from the db use same same connection under the hood when running the [recommended set up](https://developer.android.com/training/data-storage/sqlite). It may be possible to break the existing recommended pattern on Android to use distinct connections for reading and for writing from the db to gain performance which would involve some refactor for support on the SDK

**Links :** 
- [SQLite WAL](https://sqlite.org/wal.html)
- [SQLite Android dev](https://developer.android.com/training/data-storage/sqlite)
- [Android SQLite Locking](https://touchlabblog.tumblr.com/post/24474398246/android-sqlite-locking)

**Comment by pld** [comment]
Makes sense

**Pull request closed without merging**